# Human Activity Recognition with Smartphones

Refer to link https://www.kaggle.com/uciml/human-activity-recognition-with-smartphones  for more details.

The dataset is also available on UCI Machine Learning Repository.

* Problem Type : Multi-Class Classification
* Algorithm  : SVM

# Import Library

In [1]:
import numpy as np
import pandas as pd

import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline

# import SVC classifier
from sklearn.svm import SVC

# import metrics to compute accuracy (Evulate)
from sklearn.metrics import accuracy_score, confusion_matrix,classification_report
from sklearn.model_selection import cross_val_score, GridSearchCV

# Load the Training DataSet

In [2]:
df_train = pd.read_csv("archive/train.csv")

# EDA

In [3]:
df_train.head()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  tBodyAcc-std()-X  \
0           0.288585          -0.020294          -0.132905         -0.995279   
1           0.278419          -0.016411          -0.123520         -0.998245   
2           0.279653          -0.019467          -0.113462         -0.995380   
3           0.279174          -0.026201          -0.123283         -0.996091   
4           0.276629          -0.016570          -0.115362         -0.998139   

   tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  tBodyAcc-mad()-Y  \
0         -0.983111         -0.913526         -0.995112         -0.983185   
1         -0.975300         -0.960322         -0.998807         -0.974914   
2         -0.967187         -0.978944         -0.996520         -0.963668   
3         -0.983403         -0.990675         -0.997099         -0.982750   
4         -0.980817         -0.990482         -0.998321         -0.979672   

   tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  fBodyBodyGyroJerkMag-kurtosis()  \
0         -0.923527         -0.934724  ...                        -0.710304   
1         -0.957686         -0.943068  ...                        -0.861499   
2         -0.977469         -0.938692  ...                        -0.760104   
3         -0.989302         -0.938692  ...                        -0.482845   
4         -0.990441         -0.942469  ...                        -0.699205   

   angle(tBodyAccMean,gravity)  angle(tBodyAccJerkMean),gravityMean)  \
0                    -0.112754                              0.030400   
1                     0.053477                             -0.007435   
2                    -0.118559                              0.177899   
3                    -0.036788                             -0.012892   
4                     0.123320                              0.122542   

   angle(tBodyGyroMean,gravityMean)  angle(tBodyGyroJerkMean,gravityMean)  \
0                         -0.464761                             -0.018446   
1                         -0.732626                              0.703511   
2                          0.100699                              0.808529   
3                          0.640011                             -0.485366   
4                          0.693578                             -0.615971   

   angle(X,gravityMean)  angle(Y,gravityMean)  angle(Z,gravityMean)  subject  \
0             -0.841247              0.179941             -0.058627        1   
1             -0.844788              0.180289             -0.054317        1   
2             -0.848933              0.180637             -0.049118        1   
3             -0.848649              0.181935             -0.047663        1   
4             -0.847865              0.185151             -0.043892        1   

   Activity  
0  STANDING  
1  STANDING  
2  STANDING  
3  STANDING  
4  STANDING  

[5 rows x 563 columns]

In [4]:
df_train.tail()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  \
7347           0.299665          -0.057193          -0.181233   
7348           0.273853          -0.007749          -0.147468   
7349           0.273387          -0.017011          -0.045022   
7350           0.289654          -0.018843          -0.158281   
7351           0.351503          -0.012423          -0.203867   

      tBodyAcc-std()-X  tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  \
7347         -0.195387          0.039905          0.077078         -0.282301   
7348         -0.235309          0.004816          0.059280         -0.322552   
7349         -0.218218         -0.103822          0.274533         -0.304515   
7350         -0.219139         -0.111412          0.268893         -0.310487   
7351         -0.269270         -0.087212          0.177404         -0.377404   

      tBodyAcc-mad()-Y  tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  \
7347          0.043616          0.060410          0.210795  ...   
7348         -0.029456          0.080585          0.117440  ...   
7349         -0.098913          0.332584          0.043999  ...   
7350         -0.068200          0.319473          0.101702  ...   
7351         -0.038678          0.229430          0.269013  ...   

      fBodyBodyGyroJerkMag-kurtosis()  angle(tBodyAccMean,gravity)  \
7347                        -0.880324                    -0.190437   
7348                        -0.680744                     0.064907   
7349                        -0.304029                     0.052806   
7350                        -0.344314                    -0.101360   
7351                        -0.740738                    -0.280088   

      angle(tBodyAccJerkMean),gravityMean)  angle(tBodyGyroMean,gravityMean)  \
7347                              0.829718                          0.206972   
7348                              0.875679                         -0.879033   
7349                             -0.266724                          0.864404   
7350                              0.700740                          0.936674   
7351                             -0.007739                         -0.056088   

      angle(tBodyGyroJerkMean,gravityMean)  angle(X,gravityMean)  \
7347                             -0.425619             -0.791883   
7348                              0.400219             -0.771840   
7349                              0.701169             -0.779133   
7350                             -0.589479             -0.785181   
7351                             -0.616956             -0.783267   

      angle(Y,gravityMean)  angle(Z,gravityMean)  subject          Activity  
7347              0.238604              0.049819       30  WALKING_UPSTAIRS  
7348              0.252676              0.050053       30  WALKING_UPSTAIRS  
7349              0.249145              0.040811       30  WALKING_UPSTAIRS  
7350              0.246432              0.025339       30  WALKING_UPSTAIRS  
7351              0.246809              0.036695       30  WALKING_UPSTAIRS  

[5 rows x 563 columns]

In [5]:
df_train.shape

(7352, 563)

So the train data set has 563 Columns / features (including the target / class), and 7352 rows or data points.

Also the target is Activity. As mentioned in the data description it has 6 unique values. Lets check them also in next step.

## Check for missing values in the dataset

In [6]:
df_train.isnull().values.any()

False

## Class Distribution

In [7]:
df_train["Activity"].unique()

array(['STANDING', 'SITTING', 'LAYING', 'WALKING', 'WALKING_DOWNSTAIRS',
       'WALKING_UPSTAIRS'], dtype=object)

In [8]:
pd.crosstab(index = df_train["Activity"],columns="count")

col_0               count
Activity                 
LAYING               1407
SITTING              1286
STANDING             1374
WALKING              1226
WALKING_DOWNSTAIRS    986
WALKING_UPSTAIRS     1073

In [11]:
X = pd.DataFrame(df_train.drop(['Activity','subject'],axis=1))
Y = df_train.Activity.values.astype(object)

X.shape, Y.shape

((7352, 561), (7352,))

In [12]:
X.head()

tBodyAcc-mean()-X  tBodyAcc-mean()-Y  tBodyAcc-mean()-Z  tBodyAcc-std()-X  \
0           0.288585          -0.020294          -0.132905         -0.995279   
1           0.278419          -0.016411          -0.123520         -0.998245   
2           0.279653          -0.019467          -0.113462         -0.995380   
3           0.279174          -0.026201          -0.123283         -0.996091   
4           0.276629          -0.016570          -0.115362         -0.998139   

   tBodyAcc-std()-Y  tBodyAcc-std()-Z  tBodyAcc-mad()-X  tBodyAcc-mad()-Y  \
0         -0.983111         -0.913526         -0.995112         -0.983185   
1         -0.975300         -0.960322         -0.998807         -0.974914   
2         -0.967187         -0.978944         -0.996520         -0.963668   
3         -0.983403         -0.990675         -0.997099         -0.982750   
4         -0.980817         -0.990482         -0.998321         -0.979672   

   tBodyAcc-mad()-Z  tBodyAcc-max()-X  ...  fBodyBodyGyroJerkMag-meanFreq()  \
0         -0.923527         -0.934724  ...                        -0.074323   
1         -0.957686         -0.943068  ...                         0.158075   
2         -0.977469         -0.938692  ...                         0.414503   
3         -0.989302         -0.938692  ...                         0.404573   
4         -0.990441         -0.942469  ...                         0.087753   

   fBodyBodyGyroJerkMag-skewness()  fBodyBodyGyroJerkMag-kurtosis()  \
0                        -0.298676                        -0.710304   
1                        -0.595051                        -0.861499   
2                        -0.390748                        -0.760104   
3                        -0.117290                        -0.482845   
4                        -0.351471                        -0.699205   

   angle(tBodyAccMean,gravity)  angle(tBodyAccJerkMean),gravityMean)  \
0                    -0.112754                              0.030400   
1                     0.053477                             -0.007435   
2                    -0.118559                              0.177899   
3                    -0.036788                             -0.012892   
4                     0.123320                              0.122542   

   angle(tBodyGyroMean,gravityMean)  angle(tBodyGyroJerkMean,gravityMean)  \
0                         -0.464761                             -0.018446   
1                         -0.732626                              0.703511   
2                          0.100699                              0.808529   
3                          0.640011                             -0.485366   
4                          0.693578                             -0.615971   

   angle(X,gravityMean)  angle(Y,gravityMean)  angle(Z,gravityMean)  
0             -0.841247              0.179941             -0.058627  
1             -0.844788              0.180289             -0.054317  
2             -0.848933              0.180637             -0.049118  
3             -0.848649              0.181935             -0.047663  
4             -0.847865              0.185151             -0.043892  

[5 rows x 561 columns]

In [13]:
Y[1]

'STANDING'

## Transforming non numerical labels into numerical labels

In [15]:
from sklearn import preprocessing
encoder = preprocessing.LabelEncoder()

In [16]:
# encoding train labels
encoder.fit(Y)
y = encoder.transform(Y)
y.shape

(7352,)

In [17]:
y[1]

2

In [18]:
encoder.classes_

array(['LAYING', 'SITTING', 'STANDING', 'WALKING', 'WALKING_DOWNSTAIRS',
       'WALKING_UPSTAIRS'], dtype=object)

## Feature Scaling

In [20]:
# Scaling the feature
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()

In [21]:
X = scaler.fit_transform(X)
X[1]

array([ 0.05594788,  0.03148567, -0.25390836, -0.87542555, -0.92390168,
       -0.84930444, -0.86853066, -0.92199844, -0.84892814, -0.87135938,
       -0.89222832, -0.88919739,  0.89599818,  0.86865566,  0.76650295,
       -0.92069378, -0.69321044, -0.7751054 , -0.68758451, -0.84720186,
       -0.89918005, -0.82049508, -1.3048928 , -0.83019391, -1.09612841,
        2.37604697, -1.75325238,  1.32173624,  0.66992736,  0.55149006,
       -0.54554241,  0.64334534, -1.18164246,  0.26766528, -0.25635684,
       -0.22540599, -0.61013507,  0.30760857,  0.36578238,  1.23101989,
        0.5852968 , -0.40986684,  0.04445061, -0.42168323, -0.39103557,
        0.049352  , -0.42422787, -0.38623719,  0.04739076,  0.5644588 ,
       -0.43846821,  0.06881901,  0.60165647, -0.37364633,  0.06328303,
       -0.76084948,  0.67509357, -0.54965394, -0.53632953, -0.43123632,
       -0.36770227,  0.01296575, -0.91072978, -0.46503661,  1.19620895,
        0.4949101 , -0.67067971,  0.83501302, -0.9770558 ,  0.50

# Split X and y into training and validation sets

In [22]:
from sklearn.model_selection import train_test_split

X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size = 0.2, random_state = 99)
X_train.shape, X_valid.shape, y_train.shape, y_valid.shape

((5881, 561), (1471, 561), (5881,), (1471,))

# Train the Model

## Run SVM with default hyperparameters
Default hyperparameter means C=1.0, kernel=rbf and gamma=auto among other parameters.

In [23]:
# instantiate classifier with default hyperparameters
svc = SVC()

In [25]:
# fit classifier to training set
svc.fit(X_train,y_train)

SVC()

In [26]:
# make predictions on test set
y_pred = svc.predict(X_valid)

In [27]:
# compute and print accuracy score
print('Model accuracy score with default hyperparameters: {0:0.4f}'. format(accuracy_score(y_valid, y_pred)))

Model accuracy score with default hyperparameters: 0.9803


## Run SVM with rbf kernel and C=100.0

Some time there are outliers in the dataset. In that case, we should increase the value of C as higher C means fewer outliers. So, might run SVM with kernel=rbf and C=100.0.

We will try playing with various hyper-parameter.

In [28]:
# instantiate classifier with rbf kernel and C=100
svc = SVC(C=100.0)


# fit classifier to training set
svc.fit(X_train,y_train)


# make predictions on test set
y_pred = svc.predict(X_valid)


# compute and print accuracy score
print('Model accuracy score with rbf kernel and C=100.0 : {0:0.4f}'. format(accuracy_score(y_valid, y_pred)))

Model accuracy score with rbf kernel and C=100.0 : 0.9918


We can see that we obtain a higher accuracy with C=100.0 as higher C means less outliers.



## Run SVM with linear kernel
Run SVM with linear kernel and C=1.0

In [30]:
# instantiate classifier with linear kernel and C=1.0
linear_svc=SVC(kernel='linear', C=1.0)


# fit classifier to training set
linear_svc.fit(X_train,y_train)


# make predictions on test set
y_pred_test=linear_svc.predict(X_valid)


# compute and print accuracy score
print('Model accuracy score with linear kernel and C=1.0 : {0:0.4f}'. format(accuracy_score(y_valid, y_pred_test)))

Model accuracy score with linear kernel and C=1.0 : 0.9844


## Compare the train-set and test-set accuracy
Now, I will compare the train-set and test-set accuracy to check for overfitting.

In [33]:
y_pred_train = linear_svc.predict(X_train)

y_pred_train

array([5, 0, 1, ..., 2, 1, 1])

In [34]:
print('Training-set accuracy score: {0:0.4f}'. format(accuracy_score(y_train, y_pred_train)))

Training-set accuracy score: 0.9971


We can see that the training set and test-set accuracy are very much comparable.

# Check for overfitting and underfitting

In [35]:
print('Training set score: {:.4f}'.format(linear_svc.score(X_train, y_train)))

print('Validation set score: {:.4f}'.format(linear_svc.score(X_valid, y_valid)))

Training set score: 0.9971
Validation set score: 0.9844


## Run SVM with polynomial kernel and C=100.0

In [37]:
# instantiate classifier with polynomial kernel and C=100.0
poly_svc100=SVC(kernel='poly', C=100.0)


# fit classifier to training set
poly_svc100.fit(X_train, y_train)


# make predictions on test set
y_pred=poly_svc100.predict(X_valid)


# compute and print accuracy score
print('Model accuracy score with polynomial kernel and C=1.0 : {0:0.4f}'. format(accuracy_score(y_valid, y_pred)))

Model accuracy score with polynomial kernel and C=1.0 : 0.9878


## Run SVM with sigmoid kernel and C=100.0

In [39]:
# instantiate classifier with sigmoid kernel and C=100.0
sigmoid_svc100=SVC(kernel='sigmoid', C=100.0)


# fit classifier to training set
sigmoid_svc100.fit(X_train,y_train)


# make predictions on test set
y_pred=sigmoid_svc100.predict(X_valid)


# compute and print accuracy score
print('Model accuracy score with sigmoid kernel and C=100.0 : {0:0.4f}'. format(accuracy_score(y_valid, y_pred)))

Model accuracy score with sigmoid kernel and C=100.0 : 0.8477


We can see that sigmoid kernel is also performing poorly just like with polynomial kernel.

# Hyperparameter tuning using grid search and cross validation

In [42]:
# Create the parameter grid based on the results of random search
params_grid = [{'kernel': ['rbf'], 'gamma': [1e-3, 1e-4],
                     'C': [1, 10, 100, 1000]},
                    {'kernel': ['linear'], 'C': [1, 10, 100, 1000]}]

In [43]:
# Performing CV to tune parameters for best SVM fit
svm_model = GridSearchCV(SVC(), params_grid, cv=5)
svm_model.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=SVC(),
             param_grid=[{'C': [1, 10, 100, 1000], 'gamma': [0.001, 0.0001],
                          'kernel': ['rbf']},
                         {'C': [1, 10, 100, 1000], 'kernel': ['linear']}])

## Confusion Matrix and Accuracy Score

In [46]:
# Making the Confusion Matrix
#print(pd.crosstab(Y_test_label, Y_pred_label, rownames=['Actual Activity'], colnames=['Predicted Activity']))
print(confusion_matrix(y_valid,Y_pred))
print("\n")
print(classification_report(y_valid,Y_pred))

print("Training set score for SVM: %f" % final_model.score(X_train , y_train))
print("Validation set score for SVM: %f" % final_model.score(X_valid  , y_valid ))

# svm_model.score

[[249   0   0   0   0   0]
 [  0 272   5   0   0   0]
 [  0   9 284   0   0   0]
 [  0   0   0 251   0   0]
 [  0   0   0   0 186   0]
 [  0   0   0   0   0 215]]


              precision    recall  f1-score   support

           0       1.00      1.00      1.00       249
           1       0.97      0.98      0.97       277
           2       0.98      0.97      0.98       293
           3       1.00      1.00      1.00       251
           4       1.00      1.00      1.00       186
           5       1.00      1.00      1.00       215

    accuracy                           0.99      1471
   macro avg       0.99      0.99      0.99      1471
weighted avg       0.99      0.99      0.99      1471

Training set score for SVM: 1.000000
Validation set score for SVM: 0.990483
